### PR TITLE
Use custom serialization and deserialization to flatten regexFlags.

### PR DESCRIPTION
**Pull request by sima-zhu**
Working on adding tests.

### CODE REVIEW

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.search.documents.util;
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils]: This isn't marked as a Java file

Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils]: I am debugging this for a while

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+    /**
+     * Combines a list of {@link RegexFlags} into one separated by "|".
+     * E.g. {@code "CANON_EQ|CASE_INSENSITIVE|COMMENTS"}
+     *
+     * @param regexFlags A list a {@link RegexFlags} to combine.
+     * @return New combined {@link RegexFlags} separated by vertical bar.
+     */
+    public static RegexFlags combineRegexFlags(RegexFlags... regexFlags) {
+        return RegexFlags.fromString(String.join("|", Arrays.deepToString(regexFlags)));
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils]: This won't work as expected for a few reasons.

- If `regexFlags` is `null` this will return `"null"`.
- If `regexFlags` has values it will result in `"[val1.toString(), val2.toString(), ...]"`.

This would be a better option.

```java
if (CoreUtils.isNullOrEmpty(regexFlags)) {
    return null;
}

return Arrays.stream(regexFlags)
    .map(RegexFlags::toString)
    .collect(Collectors.joining("|"));
```

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+public final class RegexFlagsUtils {
+
+    /**
+     * Combines a list of {@link RegexFlags} into one separated by "|".
+     * E.g. {@code "CANON_EQ|CASE_INSENSITIVE|COMMENTS"}
+     *
+     * @param regexFlags A list a {@link RegexFlags} to combine.
+     * @return New combined {@link RegexFlags} separated by vertical bar.
+     */
+    public static RegexFlags combineRegexFlags(RegexFlags... regexFlags) {
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils]: I'm a bit worried that `RegexFlags` is an `ExpandableStringEnum` as this may grow unbounded with the number of permutations available and will remain in memory as it is backed by a static dictionary. 

@brjohnstmsft does the ordering of the regex flags matter? If not we should sort before joining as that reduces the number of possible combinations that will be stored in static memory.

Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils]: Would the service return a combined regex flag that a user may want to split up? If so should we offer a `splitRegexFlags` class that has this API shape:

```java
public static RegexFlags[] splitRegexFlags(RegexFlags combinedRegexFlag)
```

File: sdk/search/azure-search/src/test/java/com/azure/search/documents/CustomAnalyzerSyncTests.java [hunk 4]
```diff
@@ -313,6 +314,20 @@ public void canUseAllAnalysisComponentNames() {
         assertCustomAnalysisComponentsEqual(index, createdIndex);
     }
 
+    @Test
+    public void canUseCombinedRegexFlags() {
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/test/java/com/azure/search/documents/CustomAnalyzerSyncTests.java]: Should add some tests that target `RegexFlagsUtils` directly.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils (hunk 3 above)
Comment (brjohnstmsft) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils]: The ordering shouldn't matter, but maybe we should consider modeling the `flags` properties of `PatternAnalyzer` and `PatternTokenizer` as collections instead of introducing this combining primitive? I believe @xirzec is taking this approach in TypeScript.

Comment (brjohnstmsft) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils]: Nobody has asked for this, but the scenario makes sense. It seems like one more reason to just model the appropriate properties as lists in the first place.

Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils]: I have considered to do this. These two classes are final auto-gen classes, we have to do some transformation work if adding this function over there.

Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils]: Sync with @xirzec , I will move this function to the best place after discussing with Jeff.

Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils]: This is nice to have feature and addon. I would prefer to have it after GA

**Comment by sima-zhu** [comment]
Postpone the PR a bit after I synced with other languages.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils (hunk 3 above)
Comment (brjohnstmsft) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/util/RegexFlagsUtils]: @sima-zhu If you model flags as string collections, you'll need to implement the splitting logic for deserialization anyway. This should be done sooner rather than later.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/models/PatternAnalyzer.java [hunk 5]
```diff
@@ -112,8 +114,9 @@ public RegexFlags getFlags() {
      * @param flags the flags value to set.
      * @return the PatternAnalyzer object itself.
      */
-    public PatternAnalyzer setFlags(RegexFlags flags) {
-        this.flags = flags;
+    public PatternAnalyzer setFlags(Collection<RegexFlags> flags) {
+        String newRegexFlags = flags.stream().map(RegexFlags::toString).collect(Collectors.joining("|"));
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/models/PatternAnalyzer.java]: This won't be `null` safe which is different behavior than before.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/models/PatternAnalyzer.java [hunk 6]
```diff
@@ -112,8 +114,9 @@ public RegexFlags getFlags() {
      * @param flags the flags value to set.
      * @return the PatternAnalyzer object itself.
      */
-    public PatternAnalyzer setFlags(RegexFlags flags) {
-        this.flags = flags;
+    public PatternAnalyzer setFlags(Collection<RegexFlags> flags) {
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/models/PatternAnalyzer.java]: This API may be better as `setFlags(RegexFlags... flags)`.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/models/PatternAnalyzer.java (hunk 5 above)
Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/models/PatternAnalyzer.java]: Yes, it will throw NPE. Is it better to make it silent if pass null?

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/models/PatternAnalyzer.java (hunk 6 above)
Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/models/PatternAnalyzer.java]: The reason I choose this because we do not need to convert if passing RegexFlags.values()

Comment (alzimmermsft) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/models/PatternAnalyzer.java]: Would this be a common scenario though? Also, I think this would be likely to fail as `values()` would contain all the unioned `RegexFlags` as well.

Comment (alzimmermsft) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/models/PatternAnalyzer.java]: @brjohnstmsft what does the service do if the same flag is passed multiple times.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/models/PatternAnalyzer.java (hunk 5 above)
Comment (alzimmermsft) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/models/PatternAnalyzer.java]: `null` was an allowable configuration before, this changes it so that once this is set you can never clear it out.

File: sdk/search/azure-search/src/main/java/module-info.java [hunk 7]
```diff
@@ -5,12 +5,14 @@
     requires transitive com.azure.core;
     requires com.fasterxml.jackson.datatype.jsr310;
     requires jakarta.activation;
+    requires com.fasterxml.jackson.dataformat.xml;
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/module-info.java]: Why are we requiring `dataformat.xml`?

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerSerializer.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+     * {@inheritDoc}
+     *
+     * @param analyzer The {@link PatternAnalyzer} needs to serialize
+     * @param jsonGenerator Generator used to output resulting Json content
+     * @param serializerProvider Provider that can be used to get serializers for
+     *   serializing Objects value contains, if any.
+     * @throws IOException If IO operation error occurs.
+     */
+    @Override
+    public void serialize(final PatternAnalyzer analyzer, final JsonGenerator jsonGenerator,
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerSerializer.java]: What happens if this is called directly instead of being called through `serializeWithType`? Is it missing the start and end object pieces?

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerSerializer.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+     * @throws IOException If IO operation error occurs.
+     */
+    @Override
+    public void serializeWithType(PatternAnalyzer analyzer, JsonGenerator gen, SerializerProvider serializers,
+        TypeSerializer typeSer) throws IOException {
+        gen.writeStartObject();
+        gen.writeStringField(typeSer.getPropertyName(), typeSer.getTypeIdResolver().idFromValue(analyzer));
+        serialize(analyzer, gen, serializers);
+        gen.writeEndObject();
+    }
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerSerializer.java]: Should any logic in here just be added into `serialize` and this acts as a proxy to `serialize`.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public PatternAnalyzer deserialize(final JsonParser p, final DeserializationContext ctxt) throws IOException {
+        ObjectMapper mapper = new ObjectMapper();
+        ObjectNode root = mapper.readTree(p);
+        JsonNode flattenFlags = root.get("flags");
+        List<RegexFlags> regexFlags = Arrays.stream(flattenFlags.asText().split(DELIMITER))
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java]: What happens if `flattenFlags` is null?

Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java]: Does this need to live outside of the `fields.hasNext()` loop?

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+                analyzer.setName(field.getValue().asText());
+            } else if ("pattern".equals(field.getKey())) {
+                analyzer.setPattern(field.getValue().asText());
+            } else if ("flags".equals(field.getKey())) {
+                analyzer.setFlags(regexFlags);
+            } else if ("lowercase".equals(field.getKey())){
+                analyzer.setLowerCaseTerms(field.getValue().asBoolean());
+            } else if ("stopwords".equals(field.getKey())) {
+                List<String> stopWords = new ArrayList<>();
+                field.getValue().forEach(
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java]: What if `stopwords` is `null`?

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+                    jsonNode -> stopWords.add(jsonNode.asText())
+                );
+                analyzer.setStopwords(stopWords);
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java]: Could this be done in a stream?

```java
analyzer.setStopwords(field.getValue().stream()
    .map(node -> node.asText()).collect(Collectors.toList()));
```

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Custom deserializer for {@link PatternAnalyzer}, which convert flatten string to the list of {@link RegexFlags}.
+ */
+public class CustomPatternAnalyzerDeserializer extends JsonDeserializer<PatternAnalyzer> {
+    private static final String DELIMITER = "\\|";
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java]: Does this need to be escaped?

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerSerializer.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+     * @param serializerProvider Provider that can be used to get serializers for
+     *   serializing Objects value contains, if any.
+     * @throws IOException If IO operation error occurs.
+     */
+    @Override
+    public void serialize(final PatternAnalyzer analyzer, final JsonGenerator jsonGenerator,
+        final SerializerProvider serializerProvider) throws IOException {
+        if (analyzer == null) {
+            return;
+        }
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerSerializer.java]: What is the output when analyzer is null?

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternTokenizerDeserializer.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public PatternTokenizer deserialize(final JsonParser p, final DeserializationContext ctxt) throws IOException {
+        ObjectMapper mapper = new ObjectMapper();
+        ObjectNode root = mapper.readTree(p);
+        JsonNode flattenFlags = root.get("flags");
+        List<RegexFlags> regexFlags = Arrays.stream(flattenFlags.asText().split(DELIMITER))
+            .map(RegexFlags::fromString).collect(Collectors.toList());
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternTokenizerDeserializer.java]: Same questions as the other deserializer about `flags` being `null` and this needing to be outside of the JSON node loop.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternTokenizerDeserializer.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+        while (fields.hasNext()) {
+            Map.Entry<String, JsonNode> field = fields.next();
+            if ("name".equals(field.getKey())) {
+                tokenizer.setName(field.getValue().asText());
+            } else if ("pattern".equals(field.getKey())) {
+                tokenizer.setPattern(field.getValue().asText());
+            } else if ("flags".equals(field.getKey())) {
+                tokenizer.setFlags(regexFlags);
+            } else if ("group".equals(field.getKey())){
+                tokenizer.setGroup(field.getValue().asInt());
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternTokenizerDeserializer.java]: How does `asInt()` handle if `getValue()` is `null`? `group` is a nullable `Integer` type so it could come back as `null` from my understanding.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternTokenizerSerializer.java [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+     * @param typeSer Type serializer to use for including type information
+     * @throws IOException if IO operation error occurs.
+     */
+    @Override
+    public void serializeWithType(PatternTokenizer tokenizer, JsonGenerator gen, SerializerProvider serializers,
+        TypeSerializer typeSer) throws IOException {
+        gen.writeStartObject();
+        gen.writeStringField(typeSer.getPropertyName(), typeSer.getTypeIdResolver().idFromValue(tokenizer));
+        serialize(tokenizer, gen, serializers);
+        gen.writeEndObject();
```
Comment (alzimmermsft) [inline sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternTokenizerSerializer.java]: Same questions.

File: sdk/search/azure-search/src/main/java/module-info.java (hunk 7 above)
Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/module-info.java]: It is useless. Removed.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerSerializer.java (hunk 8 above)
Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerSerializer.java]: ["@odata.type": "#Microsoft.Azure.Search.PatternAnalyzer" ] is added before serialize the rest. So the method is called to add the polymorphism type.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerSerializer.java (hunk 9 above)
Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerSerializer.java]: Not really, the typeInfoId and typeInfoName is inside of TypeSerializer.
`["@odata.type": "#Microsoft.Azure.Search.PatternAnalyzer" ]`

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java (hunk 10 above)
Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java]: That's a miss. Will double check null cases for all.

Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java]: No. I forgot to move it into method

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java (hunk 11 above)
Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java]: Then the check of  "stopwords".equals(field.getKey()) will return false.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java (hunk 13 above)
Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java]: That's how `String.split` method works.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternTokenizerDeserializer.java (hunk 16 above)
Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternTokenizerDeserializer.java]: I don't think the case exists.
If the field value is null, then there is no need to generate field key. If the case exists for required field, then we need to have null or empty checking for the field.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerSerializer.java (hunk 14 above)
Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerSerializer.java]: > My main concern is around the serializers and deserializers and some corner cases they may have, could you add tests which specifically target them? Make sure to add guards even in scenarios where the service may not have those returns now since it is always possible to change in the future, especially around default JSON values.

Totally agreed for adding more tests to cover more corner scenarios. Will keep working on this while reviewing the body.

I have several test cases change to cover the major workflow.

**Comment by sima-zhu** [comment]
@alzimmermsft Totally agreed for adding more tests to cover more corner scenarios. Will keep working on this while reviewing the body.

I have several test cases change to cover the major workflow.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java (hunk 12 above)
Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java]: Double checked code. `field.getVaule` returns iterator.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerSerializer.java (hunk 14 above)
Comment (sima-zhu) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerSerializer.java]: I cannot have the case. `Index.setAnalyzer` takes list<Analyzer>

There is no way to have a null PatternAnalyzer pass in. 

I plan to remove it.

File: sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java (hunk 11 above)
Comment (alzimmermsft) [reply sdk/search/azure-search/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java]: So if the JSON contains the following we'll never have a `stopwords` field?

```json
{ "stopwords": null }
```

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+
+        Iterator<Map.Entry<String, JsonNode>> fields = root.fields();
+        PatternAnalyzer analyzer = new PatternAnalyzer();
+        while (fields.hasNext()) {
+            Map.Entry<String, JsonNode> field = fields.next();
+            if ("name".equals(field.getKey())) {
+                analyzer.setName(field.getValue().asText());
+            } else if ("pattern".equals(field.getKey())) {
+                analyzer.setPattern(field.getValue().asText());
+            } else if ("flags".equals(field.getKey()) && !"null".equals(field.getValue().asText())) {
```
[the remark this answers is not in the document]
Comment (sima-zhu) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/util/CustomPatternAnalyzerDeserializer.java]: "null" is a string type in json

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/util/CustomPatternTokenizerDeserializer.java [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+        ObjectNode root = mapper.readTree(p);
+        Iterator<Map.Entry<String, JsonNode>> fields = root.fields();
+        PatternTokenizer tokenizer = new PatternTokenizer();
+        while (fields.hasNext()) {
+            Map.Entry<String, JsonNode> field = fields.next();
+            if ("name".equals(field.getKey())) {
+                tokenizer.setName(field.getValue().asText());
+            } else if ("pattern".equals(field.getKey())) {
+                tokenizer.setPattern(field.getValue().asText());
+            } else if ("flags".equals(field.getKey()) && !"null".equals(field.getValue().asText())) {
```
[the remark this answers is not in the document]
Comment (sima-zhu) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/util/CustomPatternTokenizerDeserializer.java]: It is a string type in json:

`{"flags": "null"} `

what we parse out here is getValue is "null". That's why I check it like this.

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/util/CustomPatternTokenizerDeserializer.java [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+                List<RegexFlags> regexFlags = Arrays.stream(field.getValue().asText().split(DELIMITER))
+                    .map(RegexFlags::fromString).collect(Collectors.toList());
+                tokenizer.setFlags(regexFlags);
+            } else if ("group".equals(field.getKey())){
+                tokenizer.setGroup(field.getValue().asInt());
```
[the remark this answers is not in the document]
Comment (sima-zhu) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/util/CustomPatternTokenizerDeserializer.java]: Based on the test of null group, service does not return group back if it is null.
